presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java

```diff
@@ -83,6 +102,46 @@ public DruidClient(DruidConfig config, @ForDruidClient HttpClient httpClient)
         this.druidCoordinator = URI.create(config.getDruidCoordinatorUrl());
         this.druidBroker = URI.create(config.getDruidBrokerUrl());
         this.druidSchema = config.getDruidSchema();
+        this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+
+        Duration caseInsensitiveNameMatchingCacheTtl = requireNonNull(config.getCaseInsensitiveNameMatchingCacheTtl(), "caseInsensitiveNameMatchingCacheTtl is null");
+        CacheBuilder<Object, Object> remoteTableNamesCacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteTables = remoteTableNamesCacheBuilder.build();
+    }
+
+    Optional<RemoteTableObject> toRemoteTable(SchemaTableName schemaTableName)
+    {
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(schemaTableName.getTableName()), "Expected table name from internal metadata to be lowercase: %s", schemaTableName);
+        if (!caseInsensitiveNameMatching) {
+            return Optional.of(RemoteTableObject.of(schemaTableName.getTableName()));
+        }
+
+        @Nullable Optional<RemoteTableObject> remoteTable = remoteTables.getIfPresent(schemaTableName);
```

Comment:
hmmm,  can we use cache's get(K key,
    Callable<? extends V> valueLoader) to avoid returning a null?  it's something like if cached, return; otherwise create, cache and return

Reply:
@beinan In the current design, to create a new cache entry it involves looping through all the tables present in Druid. If I make use of cacheLoader to create, cache and return, it will result in looping through all the tables every time a new table is accessed as cache will be updated for only the table which is accessed. Whereas in my current implementation, if a cache reload takes place, I load all the tables into the cache which are present in Druid at the instant which means cache will only be reloaded again when a table is accessed which did not exist when the cache was refreshed the last time.

Please let me know your views on the same and correct me in case I misunderstood something.

Reply:
Got it, I'm convinced.